File: app/src/main/java/com/worldventures/dreamtrips/modules/reptools/presenter/SuccessStoryListPresenter.java
package com.worldventures.dreamtrips.modules.reptools.presenter;

import android.os.Bundle;
import android.support.v4.app.FragmentManager;

import com.techery.spares.utils.delegate.StoryLikedEventDelegate;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.modules.common.presenter.Presenter;
import com.worldventures.dreamtrips.modules.reptools.model.SuccessStory;
import com.worldventures.dreamtrips.modules.reptools.service.SuccessStoriesInteractor;
import com.worldventures.dreamtrips.modules.reptools.service.command.GetSuccessStoriesCommand;
import com.worldventures.dreamtrips.modules.reptools.view.fragment.SuccessStoryDetailsFragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.techery.janet.helper.ActionStateSubscriber;

public class SuccessStoryListPresenter extends Presenter<SuccessStoryListPresenter.View> {

   private boolean onlyFavorites = false;
   private int lastSelectedPosition = -1;

   @Inject StoryLikedEventDelegate storyLikedEventDelegate;
   @Inject SuccessStoriesInteractor successStoriesInteractor;

   @Override
   public void takeView(View view) {
      super.takeView(view);
      storyLikedEventDelegate.getObservable().compose(bindViewToMainComposer())
            .subscribe(view::updateItem);
   }

   @Override
   public void onResume() {
      if (view.getItemsCount() == 0) reload();
   }

   public void reload() {
      view.startLoading();
      successStoriesInteractor.getSuccessStoriesPipe()
            .createObservable(new GetSuccessStoriesCommand())
            .compose(bindViewToMainComposer())
            .subscribe(new ActionStateSubscriber<GetSuccessStoriesCommand>()
               .onSuccess(command -> {
                  view.finishLoading();
                  view.setItems(performFiltering(command.getResult()));
               })
               .onFail((command, throwable) -> {
                  view.finishLoading();
                  handleError(command, throwable);
               }));
   }

   @Override
   public void handleError(Object action, Throwable error) {
      super.handleError(action, error);
      view.finishLoading();
   }

   public void onSuccessStoryCellClick(SuccessStory successStory, int position) {
      handleListItemClick(successStory, position);
      view.onStoryClicked();
   }

   public void openFirst(SuccessStory successStory) {
      if (lastSelectedPosition == -1) {
         handleListItemClick(successStory, 0);
      }
   }

   public void reloadWithFilter(int filterId) {
      switch (filterId) {
         case R.id.action_show_all:
            onlyFavorites = false;
            break;
         case R.id.action_show_favorites:
            onlyFavorites = true;
            break;
      }
      reload();
   }

   public boolean isFilterFavorites() {
      return onlyFavorites;
   }

   private void handleListItemClick(SuccessStory successStory, int position) {
      lastSelectedPosition = position;
      Bundle bundle = new Bundle();
      bundle.putParcelable(SuccessStoryDetailsFragment.EXTRA_STORY, successStory);
      view.openStory(bundle);
   }

   private ArrayList<SuccessStory> performFiltering(List<SuccessStory> successStories) {
      ArrayList<SuccessStory> result = new ArrayList<>();
      if (isFilterFavorites()) {
         for (SuccessStory successStory : successStories) {
            if (successStory.isLiked()) {
               result.add(successStory);
            }
         }
      } else {
         result.addAll(successStories);
      }

      Collections.sort(result, (lhs, rhs) -> lhs.getAuthor().compareTo(rhs.getAuthor()));
      Collections.sort(result, (lhs, rhs) -> lhs.getCategory().compareTo(rhs.getCategory()));

      return result;
   }

   public interface View extends Presenter.View {

      void setItems(List<SuccessStory> items);

      void updateItem(SuccessStory item);

      int getItemsCount();

      void finishLoading();

      void startLoading();

      void onStoryClicked();

      void openStory(Bundle bundle);

      FragmentManager getSupportFragmentManager();
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/reptools/presenter/SuccessStoryListPresenter.kt
package com.worldventures.dreamtrips.social.ui.reptools.presenter

import android.os.Bundle
import com.worldventures.dreamtrips.modules.common.presenter.Presenter
import com.worldventures.dreamtrips.social.service.reptools.SuccessStoriesInteractor
import com.worldventures.dreamtrips.social.service.reptools.command.GetSuccessStoriesCommand
import com.worldventures.dreamtrips.social.service.reptools.command.ReadSuccessStoriesCommand
import com.worldventures.dreamtrips.social.ui.reptools.model.SuccessStory
import com.worldventures.dreamtrips.social.ui.reptools.service.analytics.FilterSuccessStoriesShowAllAction
import com.worldventures.dreamtrips.social.ui.reptools.service.analytics.FilterSuccessStoriesShowFavoriteAction
import com.worldventures.dreamtrips.social.ui.reptools.service.analytics.SearchSuccessStoriesAction
import com.worldventures.dreamtrips.social.ui.reptools.view.fragment.SuccessStoryDetailsFragment
import io.techery.janet.helper.ActionStateSubscriber
import javax.inject.Inject

open class SuccessStoryListPresenter : Presenter<SuccessStoryListPresenter.View>() {

   private var isFilterFavorites = false
   private var lastSelectedPosition = -1

   @Inject lateinit var successStoriesInteractor: SuccessStoriesInteractor

   override fun onViewTaken() {
      super.onViewTaken()
      subscribeSuccessStories()
      reload()
   }

   fun subscribeSuccessStories() {
      successStoriesInteractor.successStoriesPipe
            .observeSuccess()
            .compose(bindViewToMainComposer())
            .subscribe { view.setItems(it.filteredResult()) }
   }

   fun reload() {
      successStoriesInteractor.getSuccessStoriesPipe
            .createObservable(GetSuccessStoriesCommand())
            .compose(bindViewToMainComposer())
            .subscribe(ActionStateSubscriber<GetSuccessStoriesCommand>()
                  .onStart { view.startLoading() }
                  .onFinish { view.finishLoading() }
                  .onFail(this::handleError))
   }

   fun filterByQuery(query: String) {
      successStoriesInteractor.successStoriesPipe.send(ReadSuccessStoriesCommand {
         it.title.contains(query, ignoreCase = true) && (!isFilterFavorites || it.isLiked)
      })
   }

   fun filterFavorites(filterFavorites: Boolean) {
      this.isFilterFavorites = filterFavorites
      trackFilterPressed(filterFavorites)
      successStoriesInteractor.successStoriesPipe.send(ReadSuccessStoriesCommand { !isFilterFavorites || it.isLiked })
   }

   private fun trackFilterPressed(filterFavorites: Boolean) {
      if (filterFavorites) {
         analyticsInteractor.analyticsActionPipe().send(FilterSuccessStoriesShowFavoriteAction())
      } else {
         analyticsInteractor.analyticsActionPipe().send(FilterSuccessStoriesShowAllAction())
      }
   }

   fun onShowFilterRequired() = view.showFilterDialog(isFilterFavorites)

   fun onSuccessStoryCellClick(successStory: SuccessStory, position: Int) = openStory(successStory, position)

   fun openFirst(successStory: SuccessStory) {
      if (lastSelectedPosition == -1) {
         openStory(successStory, 0)
      }
   }

   fun onSearchActivated() = analyticsInteractor.analyticsActionPipe().send(SearchSuccessStoriesAction())

   fun openStory(successStory: SuccessStory, position: Int) {
      lastSelectedPosition = position
      view.openStory(Bundle().apply { putParcelable(SuccessStoryDetailsFragment.EXTRA_STORY, successStory) })
   }

   interface View : Presenter.View {
      fun setItems(items: List<SuccessStory>?)

      fun finishLoading()

      fun startLoading()

      fun openStory(bundle: Bundle)

      fun showFilterDialog(showFavorites: Boolean)
   }
}

